src/shared_utilities/implicit_timestepping.jl

```diff
@@ -0,0 +1,166 @@
+    function tendency_jacobian!(W, Y, p, dtγ, t)
+        update_aux!(p, Y, t)
+        update_boundary_fluxes!(p, Y, t)
+        update_jacobian!(W, Y, p, dtγ, t)
+    end
+    return tendency_jacobian!
+end
+
+"""
+    make_update_jacobian(model::AbstractModel)
+
+Creates and returns a function which updates the entries
+of the Jacobian matrix `W` in place.
+
+If the implicit tendency function is given by
+`T!(dY, Y, p, t) = make_implicit_tendency(model)`, the Jacobian
+should be given by `W_{i,j}! = ∂T!_i/∂Y_j`, where `Y_j` is the
+`j-th` state variable
+and `T!_i` is the implicit tendency of the `i-th` state variable.
+
+The default is that no updates are required, no implicit tendency is
+present, and hence the timestepping is entirely explicit.
+"""
+function make_update_jacobian(model::AbstractModel)
+    function update_jacobian!(W, Y, p, dtγ, t) end
+    return update_jacobian!
+end
+
+"""
+    set_dfluxBCdY!(::AbstractModel,
+                  ::AbstractBC,
+                  ::AbstractBoundary,
+                  _...)::Union{ClimaCore.Fields.FieldVector, Nothing}
+
+A function stub which returns the derivative of the implicit tendency
+term of the `model` arising from the boundary condition,
+with respect to the state Y.
+"""
+function set_dfluxBCdY!(
+    ::AbstractModel,
+    ::AbstractBC,
+    ::AbstractBoundary,
+    _...,
+)::Union{ClimaCore.Fields.FieldVector, Nothing} end
+
+"""
+    ImplicitEquationJacobian{M, S}
+
+A struct containing the necessary information for constructing a tridiagonal
+Jacobian matrix (`W`) for solving Richards equation, treating only the vertical
+diffusion term implicitly.
+
+`matrix` is a block matrix containing the tri-diagonal matrix `∂ϑres∂ϑ`
+in the RichardsModel case.
+`solver` is a diagonal solver for the RichardsModel case because our matrix is
+block diagonal.
+
+Note that the diagonal, upper diagonal, and lower diagonal entry values
+are stored in this struct and updated in place.
+"""
+struct ImplicitEquationJacobian{M, S}
+    "Jacobian matrix stored as a MatrixFields.FieldMatrix"
+    matrix::M
+    "Solver to use for solving the tridiagonal system"
+    solver::S
+end
+
+"""
+    ImplicitEquationJacobian(
+        Y::ClimaCore.Fields.FieldVector;
+)
+
+Outer constructor for the ImplicitEquationJacobian Jacobian
+matrix struct.
+
+Initializes all variables to zeros.
+"""
+function ImplicitEquationJacobian(Y::ClimaCore.Fields.FieldVector)
+    FT = eltype(Y)
+    center_space = axes(Y.soil.ϑ_l)
+
+    # Construct a tridiagonal matrix that will be used as the Jacobian
+    tridiag_type = MatrixFields.TridiagonalMatrixRow{FT}
+    # Create a field containing a `TridiagonalMatrixRow` at each point
+    tridiag_field = Fields.Field(tridiag_type, center_space)
+    fill!(parent(tridiag_field), NaN)
+
+    # Only add jacobian blocks for fields that are in Y
+    is_in_Y(name) = MatrixFields.has_field(Y, name)
+
+    implicit_names = (@name(soil.ϑ_l),)
+    explicit_names = (
```

Comment:
we mentioned opening an issue to generalize this.

Reply:
opened here: https://github.com/CliMA/ClimaLand.jl/issues/661